File: java/client/src/main/java/glide/api/models/configuration/PubSubState.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.configuration;

import glide.api.models.configuration.BaseSubscriptionConfiguration.ChannelMode;
import java.util.Map;
import java.util.Set;

/**
 * Represents the subscription state for a pubsub client.
 *
 * @param <T> The channel mode type (e.g., PubSubChannelMode or PubSubClusterChannelMode)
 */
public interface PubSubState<T extends ChannelMode> {
    /**
     * Gets the desired subscription state.
     *
     * @return Map of channel modes to their subscribed channels
     */
    Map<T, Set<String>> getDesiredSubscriptions();

    /**
     * Gets the actual subscription state.
     *
     * @return Map of channel modes to their subscribed channels
     */
    Map<T, Set<String>> getActualSubscriptions();
}

File: go/models/pubsub_state.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package models

// PubSubChannelMode represents the type of pubsub subscription.
type PubSubChannelMode int

const (
	Exact PubSubChannelMode = iota
	Pattern
	Sharded
)

// PubSubState represents the subscription state of a client.
type PubSubState struct {
	// DesiredSubscriptions contains the channels/patterns the client intends to be subscribed to.
	DesiredSubscriptions map[PubSubChannelMode]map[string]struct{}
	// ActualSubscriptions contains the channels/patterns the client is actually subscribed to on the server.
	ActualSubscriptions map[PubSubChannelMode]map[string]struct{}
}

// NewPubSubState creates a new PubSubState with initialized maps.
func NewPubSubState() *PubSubState {
	return &PubSubState{
		DesiredSubscriptions: make(map[PubSubChannelMode]map[string]struct{}),
		ActualSubscriptions:  make(map[PubSubChannelMode]map[string]struct{}),
	}
}
